### PR TITLE
Refactor output layer

### PR DESCRIPTION
**Pull request by uduse**
Refactor `tasks.BaseTask.make_output_layer` into `engine.BaseModel._make_output_layer`, so models can customize their output layers based on different task by overriding the default behavior.

### CODE REVIEW

File: matchzoo/engine/base_model.py [hunk 1]
```diff
@@ -235,6 +235,8 @@ def _make_output_layer(self):
             return keras.layers.Dense(task.num_classes, activation='softmax')
         elif isinstance(task, tasks.Ranking):
             return keras.layers.Dense(1, activation='sigmoid')
```
Comment (faneshion) [inline matchzoo/engine/base_model.py]: It's wonderful to bind the output with the task type. Here, I think the `sigmoid` is optimal in `tasks.Ranking`, since not all models require the output to be a probability. Sometimes, it is more efficient to train the model when the output to be a real number (e.g., in hinge loss).

Comment (uduse) [reply matchzoo/engine/base_model.py]: The point of moving layer constructions from tasks to models is to allow models to use different losses for the same task by overriding the default `_make_output_layer`. In your case where a model wants a different output activation function, the model will need to override `_make_output_layer` and handle `tasks.Ranking` output layer construction itself.

Comment (faneshion) [reply matchzoo/engine/base_model.py]: Well, it's well enough in current `_make_output_layer` since it can cover the majority of ranking cases. Special requirement can be satisfied by overriding.

File: matchzoo/models/dense_baseline_model.py [hunk 2]
```diff
@@ -38,5 +38,5 @@ def build(self):
         x = keras.layers.concatenate(x_in)
         x = keras.layers.Dense(
                 self._params['num_dense_units'], activation='relu')(x)
-        x_out = self.params['task'].make_output_layer()(x)
+        x_out = self._make_output_layer()(x)
```
Comment (bwanglzu) [inline matchzoo/models/dense_baseline_model.py]: why use private method instead of a public one?

File: matchzoo/models/naive_model.py [hunk 3]
```diff
@@ -13,5 +13,5 @@ def build(self):
         x_in = [keras.layers.Input(shape)
                 for shape in self._params['input_shapes']]
         x = keras.layers.concatenate(x_in)
-        x_out = self._params['task'].make_output_layer()(x)
+        x_out = self._make_output_layer()(x)
```
Comment (bwanglzu) [inline matchzoo/models/naive_model.py]: same

File: matchzoo/engine/base_model.py (hunk 1 above)
Comment (uduse) [reply matchzoo/engine/base_model.py]: @faneshion correct, that's the whole point of this PR.
